flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/RetryRule.java

```diff
@@ -70,26 +71,45 @@ public Statement apply(Statement statement, Description description) {
                     "You cannot combine the RetryOnFailure and RetryOnException annotations.");
         }
 
+        final Class<? extends Throwable> whitelistedException = getExpectedException(description);
+
         if (retryOnFailure != null) {
-            return new RetryOnFailureStatement(retryOnFailure.times(), statement);
+            return new RetryOnFailureStatement(
+                    retryOnFailure.times(), statement, whitelistedException);
```

Comment:
Hmm, I do not understand why `RetryOnFailureStatement` has `whiteListedException` as a parameter. What would be a benefit to fail early on certain exceptions?

Reply:
I would expect that `RetryOnFailureStatement` only handles assertion exceptions and throws all others if not it seems very similar to `RetryOnException`

Reply:
This is used to prevent tests annotated with `@Test(expected=IllegalArgumentException.class` to be retried multiple times; it would just waste time. The retry rule is evaluated before junit checks that the test fails as expected (=succeeded), so in this case we can exit early.
This is for example used to check various preconditions, like rejecting nulls, which would not fail with an assertion.

RetryOnException is just a special case of RetryOnFailure, limiting the retry to a specific exception that we now the test can occasionally fail with. It allows us to be a bit more selective when it comes to retries, and still have the test fail quickly if some other issue occurs.
For example, in the case of cassandra, we know the test can occasionally fail with a NoHostAvailableException, but if there is some instabilities in the produced result then we want to fail asap.

Reply:
Ah good point I did not think about `@Test(expected=IllegalArgumentException.class` case 👍